File: spendlite_web/src/lib/events.ts
export type AgentEvent =
  | { type: "delta"; text: string }
  | { type: "tool_call"; name: string; call_id: string }
  | { type: "tool_result"; name: string; call_id: string; data: unknown }
  | { type: "guardrail_triggered"; message: string }
  | { type: "done"; session_id: string }
  | { type: "error"; message: string };

export type SessionSummary = {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
};

export type HistoryMessage = { role: "user" | "assistant"; content: string };

export type CategoryRow = { category: string; total_inr: number; count: number };

export function isCategorySummary(data: unknown): data is CategoryRow[] {
  return (
    Array.isArray(data) &&
    data.length > 0 &&
    data.every(
      (r) =>
        typeof r === "object" &&
        r !== null &&
        typeof (r as CategoryRow).category === "string" &&
        typeof (r as CategoryRow).total_inr === "number",
    )
  );
}

File: spendlite_api/events.py
"""The SSE event protocol — the contract between this API and every client."""

from typing import Any, Literal

from pydantic import BaseModel


class Delta(BaseModel):
    type: Literal["delta"] = "delta"
    text: str


class ToolCall(BaseModel):
    type: Literal["tool_call"] = "tool_call"
    name: str
    call_id: str


class ToolResult(BaseModel):
    type: Literal["tool_result"] = "tool_result"
    name: str
    call_id: str
    data: Any


class GuardrailTriggered(BaseModel):
    type: Literal["guardrail_triggered"] = "guardrail_triggered"
    message: str


class Done(BaseModel):
    type: Literal["done"] = "done"
    session_id: str


class Error(BaseModel):
    type: Literal["error"] = "error"
    message: str


AgentEvent = Delta | ToolCall | ToolResult | GuardrailTriggered | Done | Error


def sse(event: AgentEvent) -> str:
    """Frame one event for the wire. The blank line terminates the frame."""
    return f"data: {event.model_dump_json()}\n\n"
